### PR TITLE
Add final shader option

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

**Comment by Meakk** [comment]
@mwestphal I will make the doc more clear but it's pretty much ready for a first pass of review.

File: _config.yml [hunk 1]
```diff
@@ -177,6 +177,13 @@ defaults:
       parent: libf3d Documentation
       nav_order: 4
 
+  -
+    scope:
+      path: "doc/libf3d/CUSTOM_SHADER.md"
```
Comment (mwestphal) [inline _config.yml:182]: if exposed in the F3D app, it should be documented in the user doc.

File: vtkext/private/module/vtkF3DUserRenderPass.cxx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  vtkOpenGLRenderWindow* renWin = static_cast<vtkOpenGLRenderWindow*>(r->GetRenderWindow());
+  vtkOpenGLState* ostate = renWin->GetState();
+
+  vtkOpenGLState::ScopedglEnableDisable bsaver(ostate, GL_BLEND);
+  vtkOpenGLState::ScopedglEnableDisable dsaver(ostate, GL_DEPTH_TEST);
+
+  if (this->DelegatePass == nullptr)
+  {
+    vtkWarningMacro("no delegate in vtkF3DUserRenderPass.");
+    return;
```
Comment (mwestphal) [inline vtkext/private/module/vtkF3DUserRenderPass.cxx:35]: can this happen ? if not, assert instead ?

File: vtkext/private/module/vtkF3DUserRenderPass.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  /**
+   * Graphics resources.
+   */
+  vtkSmartPointer<vtkOpenGLFramebufferObject> FrameBufferObject;
+  vtkSmartPointer<vtkTextureObject> ColorTexture;
+
+  std::shared_ptr<vtkOpenGLQuadHelper> QuadHelper;
+
+  std::string UserShader;
```
Comment (mwestphal) [inline vtkext/private/module/vtkF3DUserRenderPass.h:55]: put all these members as private, no reason to be protected.

File: vtkext/private/module/vtkF3DUserRenderPass.cxx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    this->FrameBufferObject->SetContext(renWin);
+  }
+
+  renWin->GetState()->PushFramebufferBindings();
+  this->RenderDelegate(s, size[0], size[1], size[0], size[1], this->FrameBufferObject, this->ColorTexture);
+  renWin->GetState()->PopFramebufferBindings();
+
+  if (this->QuadHelper && this->QuadHelper->ShaderChangeValue < this->GetMTime())
+  {
+    this->QuadHelper = nullptr;
```
Comment (mwestphal) [inline vtkext/private/module/vtkF3DUserRenderPass.cxx:65]: can this happen ? add a VTK test to cover it ?

File: _config.yml [hunk 5]
```diff
@@ -177,6 +177,13 @@ defaults:
       parent: libf3d Documentation
       nav_order: 4
 
+  -
+    scope:
+      path: "doc/libf3d/FINAL_SHADER.md"
```
Comment (mwestphal) [inline _config.yml:182]: wrong location

File: application/testing/CMakeLists.txt [hunk 6]
```diff
@@ -714,6 +714,10 @@ if(F3D_PLUGIN_BUILD_ALEMBIC AND F3D_PLUGIN_BUILD_ASSIMP)
   f3d_test(NAME TestMultiplePluginsLoad DATA cow.vtp ARGS --load-plugins=assimp,alembic NO_BASELINE REGEXP_FAIL "Plugin failed to load")
 endif()
 
+# Test opening a directory
```
Comment (mwestphal) [inline application/testing/CMakeLists.txt:717]: wrong comment

File: doc/user/FINAL_SHADER.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
+# Final shaders
+
+It is possible to specify a final full screen shader with the option `--final-shader`.
+It is executed as the last shader in the rendering pipeline.
+
+The shader code can be passed directly as a single code line.
+It's possible to pass the content of a shader file using the following command:
+
+```sh
+f3d model.obj --final-shader "$(cat final.glsl)"
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply doc/user/FINAL_SHADER.md:10]: thinking about it I'd rather have the first example to be something cross platform

File: vtkext/private/module/vtkF3DUserRenderPass.cxx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  }
+  else
+  {
+    renWin->GetShaderCache()->ReadyShaderProgram(this->QuadHelper->Program);
+  }
+
+  if (!this->QuadHelper->Program || !this->QuadHelper->Program->GetCompiled())
+  {
+    vtkErrorMacro("Couldn't build the shader program.");
+    return;
```
Comment (mwestphal) [inline vtkext/private/module/vtkF3DUserRenderPass.cxx:91]: unreachable ?

File: doc/user/FINAL_SHADER.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Meakk) [reply doc/user/FINAL_SHADER.md:10]: The whole point of this doc is to explain to the user how to use a string with new lines inside.  
I can remove it completely and let the user figure out how to do that, it's very standard after all...

File: vtkext/private/module/vtkF3DUserRenderPass.cxx (hunk 8 above)
Comment (Meakk) [reply vtkext/private/module/vtkF3DUserRenderPass.cxx:91]: covered with the new test

File: doc/user/FINAL_SHADER.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mwestphal) [reply doc/user/FINAL_SHADER.md:10]: Got it. Mention that this example may not work on Windows at least.

**Review by mwestphal — APPROVED** [review]

File: doc/user/FINAL_SHADER.md
Comment (mwestphal) [inline doc/user/FINAL_SHADER.md:1]: missing from doc/user/README_USER.md

**Review by mwestphal — CHANGES_REQUESTED** [review]
small doc change

**Review by mwestphal — APPROVED** [review]

**Pull request merged**